File: Anul_2/lab10/lab10/repository/AbstractRepository.cs
using lab10.entities;
using lab10.validation;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace lab10.repository
{
    abstract class AbstractRepository<ID,E>:IRepository<ID,E> where E:Entity<ID>
    {
        private AbstractValidator<ID, E> validator;
        private List<E> EList=new List<E>();
        public AbstractRepository(AbstractValidator<ID,E> validator)
        {
            this.validator = validator;
            //LoadFromFile();
        }
        public abstract void LoadFromFile();
        public abstract void WriteToFile();
        public void StoreFromFile(E entity)
        {
            if (entity == null)
                throw new ValidationException("Entitate vida");
            if (FindOne(entity.Id) == null)
            {
                validator.Validate(entity);
                EList.Add(entity);
            }
        }
        public E FindOne(ID id)
        {
            if (id == null)
            {
                throw new ValidationException("Id null");
            }
            var rez = EList.Where(e => e.Id.Equals(id));
            if (rez.Count() == 0)
                return null;
            return rez.First();
        }

        public List<E> FindAll()
        {
            return EList;
        }

        public E Save(E entity)
        {
            if (entity == null)
                throw new ValidationException("Entitate vida");
            if (FindOne(entity.Id) != null)
                return FindOne(entity.Id);
            validator.Validate(entity);
            EList.Add(entity);
            WriteToFile();
            return null;
        }

        public E Delete(ID id)
        {
            if (id == null)
            {
                throw new ValidationException("Id null");
            }
            E entity = FindOne(id);
            if (entity == null)
                return null;
            EList.Remove(entity);
            WriteToFile();
            return entity;
        }

        public E Update(E entity)
        {
            if (entity == null)
                throw new ValidationException("Entitate null");
            E e = FindOne(entity.Id);
            if (e == null)
                return entity;
            EList.Remove(e);
            EList.Add(entity);
            return null;
        }
    }
}

File: Anul_2/PracticJAVA/src/main/java/Repositories/AbstractRepository.java
package Repositories;

import Entities.HasID;

import java.io.*;
import java.nio.Buffer;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public abstract class AbstractRepository<ID, E extends HasID<ID>> implements Repository<ID, E> {

    private Map<ID, E> entities;
    private String fileName;

    public abstract void loadData() throws IOException;
    public abstract void writeToFile() throws FileNotFoundException;

    public AbstractRepository(String filename) throws FileNotFoundException {
        entities = new HashMap<ID, E>();
        fileName = filename;
        loadData();
    }

    @Override
    public E findOne(ID id){

        if(!entities.containsKey(id)) return null;
        return entities.get(id);
    }

    @Override
    public List<E> findAll(){

        List<E> l = new ArrayList<>();
        l.addAll(entities.values());
        return l;
    }

    @Override
    public E save(E entity) throws FileNotFoundException {
        if(this.findOne(entity.getID()) != null) return entity;
        if(this.findOne(entity.getID()) != null && entity.getID() == this.findOne(entity.getID()).getID()) return entity;
        entities.put(entity.getID(),entity);
        writeToFile();
        return null;
    }

    @Override
    public E delete(ID id) throws FileNotFoundException {
        E entity = this.findOne(id);
        entities.remove(id);
        writeToFile();
        return entity;
    }

    @Override
    public E update(E entity) throws FileNotFoundException {
        if (findOne(entity.getID()) == null) {
            return null;
        } else {
            E ret = entities.replace(entity.getID(), entity);
            writeToFile();
            return ret;
        }
    }

    @Override
    public long size(){
        return entities.size();
    }

}
